### DISCUSSION TITLE
SSL cert setup

### DISCUSSION THREAD

**Discussion by proton1k** [vercel/next.js#13546] [Help]
Hi!
Long story short: I had to use a self-signed certificate to run the app on a localhost. But I receive the following error message:

```
event - compiled successfully
Error: unable to verify the first certificate
    at TLSSocket.onConnectSecure (_tls_wrap.js:1486:34)
    at TLSSocket.emit (events.js:315:20)
    at TLSSocket._finishInit (_tls_wrap.js:921:8)
    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:695:12) {
  code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE',
```

the `/etc/hosts`, traefik config file has correct entries and I also have the `$NODE_EXTRA_CA_CERTS` pointing to the correct pem file. Still after the nextjs app started I see an error message

```
Server Error
Error: unable to verify the first certificate
```

What can be wrong here? The application also doe not launch correctly on the localhost:$PORT either.

**update**

Yes, I've checked also documentation for how the Next js handles environment vars and I've tried two things: one is adding `NODE_EXTRA_CA_CERTS="~/Library/Application Support/mkcert/rootCA.pem"` which loaded the var into environment where the application is running, but it did not remove the error message. The second one which I do not like though it's weirdly suggested everywhere - to disable the certificate verification so that any TLS / HTTPS is actually becomes insecure. Sometimes it may be ok, but not in all cases. That was done by adding the `NODE_TLS_REJECT_UNAUTHORIZED='0'` into the same `.env.local`.

**reading list:**

 + https://unix.stackexchange.com/a/541092/57585
 + https://stackoverflow.com/questions/20082893/unable-to-verify-leaf-signature
 + https://superuser.com/a/892248/167792
 + https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables
 + https://github.com/facebook/create-react-app/issues/3766

**Comment by Uranbold** [comment]
I can't resolve this? Did you resolve this issue?

**Reply by proton1k** [reply]
Nope

**Reply by kyds3k** [reply]
161 days later - still no resolution?

**Reply by Rutvij07** [reply]
Facing same issue.

**Comment by modbp** [comment]
Perhaps something has changed in the past year? Setting the `NODE_EXTRA_CA_CERTS` env var worked when I encountered this error.

In my case I am using `mkcert`, so I passed in `NODE_EXTRA_CA_CERTS="$(mkcert -CAROOT)/rootCA.pem"` as part of my npm start script. Found in [the mkcert docs](https://github.com/FiloSottile/mkcert#using-the-root-with-nodejs).

**Reply by uehlbran** [reply]
Would be nice to see how you set this up for those of use that might not be familiar with node. Especially since there is nothing on the internet about how to do this in nextjs... and everything I've tried has failed.

It appears the only way to get this to work is with cross-env as .env, .env.local and so on do not work... Thanks NextJs... Production grade framework alright.

"cross-env NODE_EXTRA_CA_CERTS=\"C:\\Users\\<username>\\AppData\\Local\\mkcert\\rootCA.pem\" next dev"

**Reply by travhow99** [reply]
This worked perfectly fine for me just using it in the npm script like so:
`"dev": "export NODE_EXTRA_CA_CERTS=\"$(mkcert -CAROOT)/rootCA.pem\" && next dev -p 3001",`

**Comment by uehlbran** [comment]
Why is a basic feature like this not supported in "A Production grade framework" out of the box?

**Reply by RightFootConsulting** [reply]
Because NextJS is not a production grade framework, NodeJS and all things NodeJS are nothing but hacks.  This is especially true for NextJS.  Great for little toy projects and youtubers, but I would not build an enterprise system on this platform.  IT's great for a prototype but when you want a reliable system better to consider your investment in NextJS a sunk cost.  If you are lookng for something wiith NextJS ease of use, and based on a platform that can be counted on for more than just toys, try something in the Golang ecosystem like Pushup.

**Reply by GabenGar** [reply]
Golang is even more of a toy language, nice try.

**Comment by GabenGar** [comment]
Not sure if relevant, but I got the https version of NextJS running locally with self-signed certificate: https://gist.github.com/GabenGar/078de497f181742fb9a669dd8dcfa130

**Comment by nedcerneckis** [comment]
Has anyone been able to solve this?

**Comment by michalprikryl** [comment]
This solution is not about nextjs (even I am developing nextjs app), but this error get me there, so I post it also here. This problem occurs for me few days ago - nodejs cannot validate certificates. After few hours of searching and reinstalling everything across node I realize that ESET Antivirus is "validating/changing" all SSL certificates. So I disable this feature (https://support.eset.com/en/kb3126-disable-ssl-filtering-in-eset-windows-products) and everything was fine. I hope that this will help someone.

**Reply by manoloramos** [reply]
@michalprikryl I've faced the same issue for 3 days now, spent hours looking for solutions and removing Node & NPM completely from my system. After trying everything I could, my suspicions began arousing around my antivirus (ESET), but I didn't take a deep look until I found your post. Everything works now, albeit there must be some solution to avoid fully deactivating the feature.

Thank you.

**Comment by tiendq** [comment]
Unanswered after 4 years for a likely popular dev issue :D

What I'm experiencing now:

`NODE_EXTRA_CA_CERTS="$(mkcert -CAROOT)/rootCA.pem" next dev` doesn't work because it requires `sudo` priviledge. You will get this warning `Warning: Ignoring extra certs from `~/Library/Application Support/mkcert/rootCA.pem`, load failed: error:80000002:system library::No such file or directory`

If you use an exact `.pem` file e.g. in my case `NODE_EXTRA_CA_CERTS=./test.pem next dev` there is no warning like above one but nothing fixed, still got `UNABLE_TO_VERIFY_LEAF_SIGNATURE` error.

I'd like to add more details for my case, I have a NextJS app running at `localhost:3000` as usual but it has `Image` which load images from different `https` app (used local self signed certificate with `mkcert`) then I got the error.

**Comment by akshaykumar90** [comment]
Using the `--experimental-https` flag with  `--experimental-https-ca` set to the rootCA path worked for me. Note that you would also have to specify `--experimental-https-key` and `--experimental-https-cert` options.

Example:

```
next dev --experimental-https --experimental-https-ca rootCA.pem --experimental-https-key key.pem --experimental-https-cert cert.pem
```

Docs: https://nextjs.org/docs/pages/api-reference/cli/next#using-https-during-development

**Reply by isarmstrong** [reply]
This is the way.
package.json
`{
  "scripts": {
    "dev": "next dev --experimental-https --experimental-https-ca rootCA.pem --experimental-https-key key.pem --experimental-https-cert cert.pem",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    // ... other scripts remain unchanged
  },
  // ... rest of package.json remains unchanged
}`

bash
`# Generate a private key
openssl genrsa -out key.pem 2048

# Generate a CSR (Certificate Signing Request)
openssl req -new -key key.pem -out csr.pem

# Generate a self-signed certificate
openssl x509 -req -days 365 -in csr.pem -signkey key.pem -out cert.pem

# Generate a root CA (Certificate Authority)
openssl genrsa -out rootCA.key 2048
openssl req -x509 -new -nodes -key rootCA.key -sha256 -days 1024 -out rootCA.pem`

Done and dusted

**Comment by dufrein** [comment]
using script "dev": "next dev --experimental-https" is enough for my case (api on expressjs, front on nextjs)
